client/blocks/user-mentions/with-user-mentions.jsx

```diff
@@ -18,33 +22,250 @@ import React from 'react';
  * @returns {object} the enhanced component
  */
 export default EnhancedComponent =>
-	class withUserMentionSuggestions extends React.Component {
-		static displayName = `withUserMentionSuggestions( ${ EnhancedComponent.displayName ||
+	class withUserMentions extends React.Component {
+		matchingSuggestions = [];
+
+		static displayName = `withUserMentions( ${ EnhancedComponent.displayName ||
 			EnhancedComponent.name } )`;
 		static propTypes = {};
 
 		state = {
 			showPopover: false,
+			popoverContext: null,
+			popoverPosition: null,
+			query: '',
 		};
 
-		handleKeyPress = e => {
-			if ( e.target.value[ e.target.value.length - 1 ] === '@' ) {
-				console.log( 'found @something' ); // eslint-disable-line no-console
-				this.setState( { showPopover: true } );
+		constructor( props ) {
+			super( props );
+			// create a ref to store the textarea DOM element
+			this.textInput = React.createRef();
+		}
+
+		componentDidMount() {
+			if ( typeof window !== 'undefined' ) {
+				window.addEventListener( 'resize', this.throttledUpdatePosition );
+			}
+		}
+
+		componentWillUpdate( nextProps, nextState ) {
+			// Update position of popover if going from invisible to visible state.
+			if ( ! this.state.showPopover && nextState.showPopover ) {
+				this.updatePosition( nextState );
+				return;
+			}
+
+			// Update position of popover if cursor has moved to a new line.
+			if ( nextState.showPopover ) {
+				const currentTop = this.state.popoverPosition && this.state.popoverPosition.top;
+				const currentLeft = this.state.popoverPosition && this.state.popoverPosition.left;
+
+				if ( currentTop && currentLeft ) {
+					const { top, left } = this.getPosition();
+					const isLineBefore = currentTop > top && currentTop < left;
+					const isLineAfter = currentTop < top && currentLeft > left;
+
+					if ( isLineBefore || isLineAfter ) {
+						this.updatePosition( nextState, { top, left } );
+					}
+				}
+			}
+		}
+
+		componentWillUnmount() {
+			if ( typeof window !== 'undefined' ) {
+				window.removeEventListener( 'resize', this.throttledUpdatePosition );
+			}
+		}
+
+		handleKeyDown = event => {
+			const selectedIndex = this.getSelectedSuggestionIndex();
+
+			if ( ! includes( [ keys.upArrow, keys.downArrow ], event.keyCode ) || -1 === selectedIndex ) {
+				return;
 			}
+
+			let nextIndex;
+
+			// Cancel the cursor move.
+			event.preventDefault();
+
+			if ( event.keyCode === keys.downArrow ) {
+				nextIndex = ( selectedIndex + 1 ) % this.matchingSuggestions.length;
+			} else {
+				nextIndex = selectedIndex - 1;
+
+				if ( nextIndex < 0 ) {
+					nextIndex = this.matchingSuggestions.length - 1;
+				}
+			}
+
+			this.setState( { selectedSuggestionId: this.matchingSuggestions[ nextIndex ].ID } );
 		};
 
+		handleKeyUp = event => {
```

Comment:
I'm unfamiliar with how this works - why do we handle user input in both `handleKeyDown` and `handleKeyUp`? Would handling it all in `handleKeyUp` lead to unexpected behavior?

Reply:
`keyDown` can repeat on a long press, but `keyUp` only ever fires once, after the key has been released:

https://javascript.info/keyboard-events

If you long press an up/down arrow, we'd want it to keep scrolling through the suggestions, hence it's handled by keyDown 🎹